Update Glossary, pt. 2 (#659)
substrate-developer-hub/substrate-developer-hub.github.io: docs/knowledgebase/getting-started/glossary.md

The loss of a system service due to a Byzantine Fault (i.e. components in the system fail and there
is imperfect information about whether a component has failed) in systems that require consensus.

### Practical Byzantine Fault Tolerance (pBFT)

An early approach to Byzantine fault tolerance. pBFT systems tolerate Byzantine behavior from up to

> one-third of participants. The communication overhead for such systems is `O(2n²)`, where `n` is the

Reviewer: ```suggestion
one-third of participants. The communication overhead for such systems is `O(n²)`, where `n` is the
```

[a comprehensive blog series](https://polkadot.network/polkadot-consensus-part-1-introduction/) that
provides a deep dive into the consensus strategies of [the Polkadot Network](#polkadot-network).

### Hybrid Consensus

A blockchain consensus protocol that consists of independent or loosely coupled mechanisms for
[block production](#author) and [finality](#finality). This allows the chain to grow as fast as
probabilistic consensus protocols, such [Aurand](#aurand), while still maintaining the same level of

> security as [instant finality](#instant-finality) consensus protocols.

Reviewer: Also block production algorithms tend to be O(n), while finality gadgets tend to be O(n^2). Separating them allows the chain to progress with lower communication overhead while choosing an acceptable `L`, or lag to finality, to make finalization O(n^2/L) by finalizing in batches separated by `L`. Not sure if this is within the scope of the glossary or should be covered elsewhere.

## Equivocating

A type of [Byzantine](#byzantine-fault-tolerance-bft) (erroneous/malicious) behavior that involves
backing multiple mutually-exclusive options within the [consensus](#consensus) mechanism.

## Ethash

A function used by some [proof-of-work](#proof-of-work) [consensus](#consensus) systems, notably
that which is used by the Ethereum blockchain. It was developed by
[a team led by Tim Hughes](https://github.com/ethereum/ethash/graphs/contributors).

## Events

A means of recording, for the benefit of the off-chain world, that some particular [state](#state)

> transition has happened. Within the context of [FRAME](#frame), events are one of a number of

Reviewer: ```suggestion
transition happened. Within the context of [FRAME](#frame), events are one of a number of
```

composable data types that each [pallet](#pallet) may individually define. Events in FRAME are
implemented as a set of ephemeral storage items that are inspected immediately after a block has

> executed and reset during block-initialization.

Reviewer: Hmmm, it's not completely wrong, but the word ephemeral evokes the wrong sense here. It's not like you can't find them anymore after the next block, you just go back to the state between blocks `n` and `n+1` to find the events from `n`.

Author: Okay I'm just reverting back to transient and leaving it at that...let me know if you have other specific suggestions for improvement.

## Executor

A means of executing a function call in a given [runtime](#runtime) with a set of dependencies.
There are two [executor](../../knowledgebase/advanced/executor) implementations present in
Substrate, _Wasm_ and _Native_.

### Native Executor

This executor uses a natively compiled runtime embedded in the [node](#node) to execute calls. This
is a performance optimization that up-to-date nodes can take advantage of.

### Wasm Executor

> This executor uses a [Wasm](#webassembly-wasm) binary and a Wasm interpreter to execute calls. Since
> Substrate-based chains include Wasm binaries as part of their [state](#state), this is guaranteed to

Reviewer: "This" isn't clear - is it the binary or the executor that's up to date?
```suggestion
This executor uses a [Wasm](#webassembly-wasm) binary or a Wasm interpreter to execute calls. Since
Substrate-based chains include Wasm binaries as part of their [state](#state), the binary is guaranteed to
```

be up-to-date regardless of the version of the blockchain [node](#node).

## Extrinsic

A piece of data that is bundled into a [block](#block) in order to express something from the
"external" (i.e. off-chain) world. There are, broadly speaking, two types of extrinsics:
transactions, which may be
[signed](../../knowledgebase/learn-substrate/extrinsics#signed-transactions) or
[unsigned](../../knowledgebase/learn-substrate/extrinsics#unsigned-transactions), and
[inherents](../../knowledgebase/learn-substrate/extrinsics#inherents), which are inserted by
[block authors](#author).

## Existential Deposit

Within the [Balances pallet](../../knowledgebase/runtime/frame#balances), this is the minimum
balance an account may have. Accounts cannot be created with a balance less than this amount, and if

> an account's balance ever drops below this amount, then it is removed entirely.

Reviewer: May want to make a note about reference counting too. Accounts are stored in the System module, and the ED is just one factor in removing accounts.

---

## Finality

A part of [consensus](#consensus) that deals with making a progression irreversible. If a

> [block](#block) is "finalized", then any [state](#state) changes it encapsulates are irreversible

Reviewer: ```suggestion
[block](#block) is "finalized", then any [state](#state) changes it encapsulates are irreversible without a hard fork
```

and it is safe to effect any off-chain repercussions that depend on them. The consensus algorithm
_must_ guarantee that finalized blocks never need reverting.

[GRANDPA](#grandpa) is the [provable finality](#provable-finality) gadget that is used by the
[Polkadot Network](#polkadot-network).

### Instant Finality

A non-probabilistic consensus protocol that gives a guarantee of finality immediately upon block
production. These tend to be [pBFT](#practical-byzantine-fault-tolerance-pbft)-based and thus very
expensive in terms of communication requirements.

### Probabilistic Finality

In these systems, finality is expressed in terms of a probability, denoted by `p`, that a proposed
block, denoted by `B`, will remain in the canonical chain; as more blocks are produced on top of
`B`, `p` approaches 1.

### Proof-of-Finality

A piece of data that can be used to prove that a particular block is finalized.

> ### Provable Finality

Reviewer: I prefer "deterministic", where else do we use "provable"?

## Runtime

The block execution logic of a blockchain, i.e. the
[state transition function](#stf-state-transition-function). In Substrate, this is stored as a
[WebAssembly](#webassembly-wasm) binary in the [chain state](#state).

---

## SHAFT

> A synonym for [GRANDPA](#grandpa) that is an acronym for SHared Ancestry Finality Tool.

Reviewer: We still see `afg` in the logs though, right?

> A synonym for [GRANDPA](#grandpa) that is an acronym for SHared Ancestry Finality Tool.

Author: @joepetrowski & @JoshOrndorff as far as I'm concerned this is the only remaining issue...

> A synonym for [GRANDPA](#grandpa) that is an acronym for SHared Ancestry Finality Tool.

Author: I ran the node template for a sec...it finalized blocks but I didn't see `afg` in the logs.

> A synonym for [GRANDPA](#grandpa) that is an acronym for SHared Ancestry Finality Tool.

Reviewer 2: Not in the logs, but it is the debugging target. You would have to do like `substrate -lafg=trace`. Probably better the change that in Substrate than keep documenting the old name here in the glossary.

> A synonym for [GRANDPA](#grandpa) that is an acronym for SHared Ancestry Finality Tool.

Author: I removed the reference to SHAFT and took the opportunity to include a cheeky commit message as a bonus :tada:

between blocks and can be used to create new blocks as part of the state transition function. In
Substrate-based blockchains, state is stored in a [trie](#trie), a data structure that supports the
efficient creation of incremental digests. This trie is exposed to the [runtime](#runtime) as
[a simple key/value map](../../knowledgebase/advanced/storage) where both keys and values can be
arbitrary byte arrays.

## State Transition Function (STF))

The logic of a [blockchain](#blockchain) that determines how the state changes when a
[block](#block) is processed. In Substrate, this is effectively equivalent to the
[runtime](#runtime).

## Storage Items

[FRAME](#frame) primitives that provide type-safe data persistence capabilities to the
[runtime](#runtime). Learn more about storage items in the Knowledge Base article for
[runtime storage](../../knowledgebase/runtime/storage).

## Substrate

A flexible framework for building modular, efficient, and upgradeable [blockchains](#blockchain).
Substrate is written in the [Rust](https://www.rust-lang.org/) programming language and is
maintained by [Parity Technologies](https://www.parity.io/).

---

## Transaction

> A type of [extrinsic](#extrinsic) that may include a signature.

Reviewer: I think this is valuable: Because validity can be determined efficiently, transactions can	
be gossipped on the network with reasonable safety against DoS attacks, much as with Bitcoin and	
Ethereum.
